academic_observatory_workflows/workflows/orcid_telescope.py

```diff
@@ -0,0 +1,761 @@
+
+    logging.info(f"Manifest saved to {orcid_batch.manifest_file}")
+
+
+def latest_modified_record_date(manifest_file_path: Union[str, PathLike]) -> pendulum.DateTime:
+    """Reads the manifest file and finds the most recent date of modification for the records
+
+    :param manifest_file_path: the path to the manifest file
+    :return: the most recent date of modification for the records
+    """
+    with open(manifest_file_path, "r") as f:
+        reader = csv.DictReader(f)
+        modified_dates = sorted([pendulum.parse(row["updated"]) for row in reader])
+    return modified_dates[-1]
+
+
+def orcid_batch_names() -> List[str]:
+    """Create a list of all the possible ORCID directories
+
+    :return: A list of all the possible ORCID directories
+    """
+    n_1_2 = [str(i) for i in range(10)]  # 0, 1, 2, 3, 4, 5, 6, 7, 8, 9
+    n_3 = n_1_2.copy() + ["X"]  # 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, X
+    combinations = list(itertools.product(n_1_2, n_1_2, n_3))  # Creates the 000 to 99X directory structure
+    return ["".join(i) for i in combinations]
+
+
+def transform_orcid_record(record_path: str) -> Union[Dict, str]:
+    """Transform a single ORCID file/record.
+    Streams the content from the blob URI and turns this into a dictionary.
+    The xml file is turned into a dictionary, a record should have either a valid 'record' section or an 'error'section.
+    The keys of the dictionary are slightly changed so they are valid BigQuery fields.
+    If the record is valid, it is returned. If it is an error, the ORCID ID is returned.
+
+    :param download_path: The path to the file with the ORCID record.
+    :param transform_folder: The path where transformed files will be saved.
+    :return: The transformed ORCID record. If the record is an error, the ORCID ID is returned.
+    :raises KeyError: If the record is not valid - no 'error' or 'record' section found
+    :raises ValueError: If the ORCID ID does not match the file name's ID
+    """
+    # Get the orcid from the record path
+    expected_orcid = re.search(ORCID_REGEX, record_path).group(0)
+
+    with open(record_path, "rb") as f:
+        orcid_dict = xmltodict.parse(f)
+
+    try:
+        orcid_record = orcid_dict["record:record"]
+    # Some records do not have a 'record', but only 'error'. We return the path to the file in this case.
+    except KeyError:
+        orcid_record = orcid_dict["error:error"]
+        return expected_orcid
+
+    # Check that the ORCID in the file name matches the ORCID in the record
+    if not orcid_record["common:orcid-identifier"]["common:path"] == expected_orcid:
+        raise ValueError(
+            f"Expected ORCID {expected_orcid} does not match ORCID in record {orcid_record['common:orcid-identifier']['common:path']}"
+        )
+
+    # Transform the keys of the dictionary so they are valid BigQuery fields
+    orcid_record = change_keys(orcid_record)
+
+    return orcid_record
+
+
+def change_keys(obj):
```

Comment:
This type of function already exists in observatory platform. Consider using this instead:

https://github.com/The-Academic-Observatory/observatory-platform/blob/b94c968879b4bef802c0012a294b616637c68be0/observatory-platform/observatory/platform/files.py#L531C15-L531C15

Reply:
Thanks Alex. Yeah I think I initially used this function, but was experimenting with speeding up the transform step. I concluded that the way I implemented wasn't significantly faster, so I may as well change it back to use the common function. :+1: